### ISSUE TITLE
Hot Reload Time Regression

### ISSUE THREAD

**Issue by iskakaushik** [flutter/flutter#92008]
https://flutter-flutter-perf.skia.org/e/?begin=1634249677&end=1634338608&keys=X0c63f55793f83f35ae2111887a71704d&num_commits=50&request_type=1&xbaroffset=26225

Points to engine -> framework roll: https://github.com/flutter/flutter/pull/91907

**Comment by johnniwinther** [comment]
The CL has been reverted. @jensjoha is looking into why it could cause the regression.

**Comment by jensjoha** [comment]
The below is going to be long and with lots of numbers.
The TL;DR is that - from the runs I've made - Johnnis commit is not at fault.
The vm changes in https://github.com/dart-lang/sdk/commit/9f3cd7a49814ac36e85392c00a24a7c764c21852 seems to cause this. This looks to be https://dart-review.googlesource.com/c/sdk/+/216687.

The below also outlines some - it seems to me - general problems:
* The build system seems to not work properly.
* git log is different from what github shows not exactly making things harder. Am I the only one finding this weird and gets confused here?

Other observations/questions are:
* Custom build engine and non-custom build engine produce different results --- how can I make custom engine builds that are more in line with the prepackaged ones?
* It would be nice if the device-lab benchmarks had options for drilling down on specific numbers to make runs faster without having to modify the code. It would also be nice if these numbers by default included the compile time.

Details:
I've applied the following patch to flutter on revision 461a3163886533ee8a6092022f8049fdd3f313c1 to
a) make it print information about the compile time
b) make it faster to run.
:

```
diff --git a/dev/devicelab/lib/tasks/hot_mode_tests.dart b/dev/devicelab/lib/tasks/hot_mode_tests.dart
index 4618a4c0f6..543c12c7a5 100644
--- a/dev/devicelab/lib/tasks/hot_mode_tests.dart
+++ b/dev/devicelab/lib/tasks/hot_mode_tests.dart
@@ -37,9 +37,9 @@ TaskFunction createHotModeTest({String? deviceIdOverride, Map<String, String>? e
     ];
     int hotReloadCount = 0;
     late Map<String, dynamic> smallReloadData;
-    late Map<String, dynamic> mediumReloadData;
-    late Map<String, dynamic> largeReloadData;
-    late Map<String, dynamic> freshRestartReloadsData;
+    // late Map<String, dynamic> mediumReloadData;
+    // late Map<String, dynamic> largeReloadData;
+    // late Map<String, dynamic> freshRestartReloadsData;
 
 
     await inDirectory<void>(flutterDirectory, () async {
@@ -69,6 +69,7 @@ TaskFunction createHotModeTest({String? deviceIdOverride, Map<String, String>? e
             }
           });
 
+/*
           mediumReloadData = await captureReloadData(options, environment, benchmarkFile, (String line, Process process) {
             if (!line.contains('Reloaded ')) {
               return;
@@ -141,6 +142,7 @@ TaskFunction createHotModeTest({String? deviceIdOverride, Map<String, String>? e
             freshRestartReloadsData =
                 json.decode(benchmarkFile.readAsStringSync()) as Map<String, dynamic>;
           }
+*/
         });
       } finally {
         flutterFrameworkSource.writeAsStringSync(oldContents);
@@ -151,6 +153,7 @@ TaskFunction createHotModeTest({String? deviceIdOverride, Map<String, String>? e
       <String, dynamic> {
         // ignore: avoid_dynamic_calls
         'hotReloadInitialDevFSSyncMilliseconds': smallReloadData['hotReloadInitialDevFSSyncMilliseconds'][0],
+/*
         // ignore: avoid_dynamic_calls
         'hotRestartMillisecondsToFrame': smallReloadData['hotRestartMillisecondsToFrame'][0],
         // ignore: avoid_dynamic_calls
@@ -187,9 +190,11 @@ TaskFunction createHotModeTest({String? deviceIdOverride, Map<String, String>? e
         'hotReloadFlutterReassembleMillisecondsAfterLargeChange': largeReloadData['hotReloadFlutterReassembleMilliseconds'][1],
         // ignore: avoid_dynamic_calls
         'hotReloadVMReloadMillisecondsAfterLargeChange': largeReloadData['hotReloadVMReloadMilliseconds'][1],
+        */
       },
       benchmarkScoreKeys: <String>[
         'hotReloadInitialDevFSSyncMilliseconds',
+/*
         'hotRestartMillisecondsToFrame',
         'hotReloadMillisecondsToFrame',
         'hotReloadDevFSSyncMilliseconds',
@@ -208,6 +213,7 @@ TaskFunction createHotModeTest({String? deviceIdOverride, Map<String, String>? e
         'hotReloadDevFSSyncMillisecondsAfterLargeChange',
         'hotReloadFlutterReassembleMillisecondsAfterLargeChange',
         'hotReloadVMReloadMillisecondsAfterLargeChange',
+*/
       ],
     );
   };
diff --git a/packages/flutter_tools/lib/src/devfs.dart b/packages/flutter_tools/lib/src/devfs.dart
index e8670eb92d..faaea47d9b 100644
--- a/packages/flutter_tools/lib/src/devfs.dart
+++ b/packages/flutter_tools/lib/src/devfs.dart
@@ -4,6 +4,7 @@
 
 import 'dart:async';
 
+import 'package:flutter_tools/src/run_hot.dart' as LALALA;
 import 'package:package_config/package_config.dart';
 import 'package:vm_service/vm_service.dart' as vm_service;
 
@@ -612,6 +613,9 @@ class DevFS {
       checkDartPluginRegistry: true, // The entry point is assumed not to have changed.
     ).then((CompilerOutput? result) {
       compileTimer.stop();
+      if (LALALA.LALA_FIRST_RUN) {
+        print("COMPILE TIME = ${compileTimer.elapsedMilliseconds} ms");
+      }
       return result;
     });
 
diff --git a/packages/flutter_tools/lib/src/run_hot.dart b/packages/flutter_tools/lib/src/run_hot.dart
index e81bf73d64..9b70860eb1 100644
--- a/packages/flutter_tools/lib/src/run_hot.dart
+++ b/packages/flutter_tools/lib/src/run_hot.dart
@@ -80,6 +80,8 @@ class DeviceReloadReport {
   List<vm_service.ReloadReport> reports; // List has one report per Flutter view.
 }
 
+bool LALA_FIRST_RUN = false;
+
 class HotRunner extends ResidentRunner {
   HotRunner(
     List<FlutterDevice> devices, {
@@ -275,11 +277,15 @@ class HotRunner extends ResidentRunner {
     }
 
     final Stopwatch initialUpdateDevFSsTimer = Stopwatch()..start();
+    LALA_FIRST_RUN = true;
     final UpdateFSReport devfsResult = await _updateDevFS(fullRestart: true);
+    initialUpdateDevFSsTimer.stop();
+    LALA_FIRST_RUN = false;
     _addBenchmarkData(
       'hotReloadInitialDevFSSyncMilliseconds',
       initialUpdateDevFSsTimer.elapsed.inMilliseconds,
     );
+    print("NOTICE hotReloadInitialDevFSSyncMilliseconds = ${initialUpdateDevFSsTimer.elapsedMilliseconds}");
     if (!devfsResult.success) {
       return 3;
     }
@@ -310,6 +316,11 @@ class HotRunner extends ResidentRunner {
     appStartedCompleter?.complete();
 
     if (benchmarkMode) {
+      if (1+1==2) {
+        final File benchmarkOutput = globals.fs.file('hot_benchmark.json');
+        benchmarkOutput.writeAsStringSync(toPrettyJson(benchmarkData));
+        return 0;
+      }
       // Wait multiple seconds for the isolate to have fully started.
       await Future<void>.delayed(const Duration(seconds: 10));
       // We are running in benchmark mode.
diff --git a/packages/flutter_tools/lib/src/test/flutter_platform.dart b/packages/flutter_tools/lib/src/test/flutter_platform.dart
index 267d088f8c..7241ced2da 100644
--- a/packages/flutter_tools/lib/src/test/flutter_platform.dart
+++ b/packages/flutter_tools/lib/src/test/flutter_platform.dart
@@ -491,6 +491,7 @@ class FlutterPlatform extends PlatformPlugin {
       // will complete.
       // B. The test device could connect to us, in which case
       // [remoteChannelFuture] will complete.
+      final Stopwatch stopwatch = Stopwatch()..start();
       globals.printTrace('test $ourTestCount: awaiting connection to test device');
       await Future.any<void>(<Future<void>>[
         testDevice.finished,
@@ -512,7 +513,7 @@ class FlutterPlatform extends PlatformPlugin {
             remoteChannel: remoteChannel,
           );
 
-          globals.printTrace('test $ourTestCount: finished');
+          globals.printTrace('test $ourTestCount: finished in ${stopwatch.elapsedMilliseconds} ms');
           await watcher?.handleFinishedTest(testDevice);
         }()
       ]);
```

I've then in `flutter/flutter/dev/devicelab` run `ANDROID_HOME=/full/path/to/Android/Sdk FLUTTER_ENGINE=/full/path/to/flutter/engine/src dart bin/run.dart -t hot_mode_dev_cycle__benchmark --local-engine=android_debug_unopt_x64`

I've run each 5 times (sometimes 6 if there seemed to be an outlier) in an attempt to not let a single bad run ruin everything.
In the below numbers are given as `a / b` where `a` is `hotReloadInitialDevFSSyncMilliseconds` and `b` is the custom stopwatch spanning the compile time (printed in the outbut when running the command as e.g. `COMPILE TIME = 120 ms`).

with custom engine at 8034050e78101d23069091bc153e5e93309d3247 (i.e. as if no roll):
```
248 / 113
264 / 112
220 / 113
224 / 111
255 / 123
```
`hotReloadInitialDevFSSyncMilliseconds` is all over the place, but compile time is mostly centered near 110.

with custom engine at 60d0fdfa2232bfc9f3a51f79cb5e9c7139bebd43 (i.e. as if with roll):
```
228 / 111
242 / 130
242 / 120
239 / 129
222 / 119
```
Compile time mostly centered near 120.

with custom engine fc733b35c5625d4b20c3511f58ea545f28c852a4:
```
252 / 109
218 / 109
262 / 150
220 / 112
283 / 177
245 / 109
```
it's jumpy but I'll conclude it's good as most compiles are around 110.

With custom engine da2884a1c5bcc1cce2c6e726554f2036713c5c20:
```
246 / 110
218 / 110
221 / 112
206 / 102
229 / 109
```
I'll conclude it's good as most compiles are around 110.

With custom engine 32450391ac2da0713fdd2736852eeb65c95d23ae:
```
243 / 124
232 / 122
235 / 119
227 / 121
226 / 116
```
I'll conclude it's bad as most compiles are around 120

With custom engine 909b35818afdcf28ecb31df358e541965aad6b63:
```
241 / 124
241 / 136
233 / 131
232 / 122
245 / 133
```
I'll conclude it's bad as all compiles are above 120

Message from git is now:
```
909b35818afdcf28ecb31df358e541965aad6b63 is the first bad commit
commit 909b35818afdcf28ecb31df358e541965aad6b63
Author: skia-flutter-autoroll <skia-flutter-autoroll@skia.org>
Date:   Fri Oct 15 03:28:01 2021 -0400

    Roll Skia from 012f7146067a to b24bad31dc05 (3 revisions) (#29203)

 DEPS                             | 2 +-
 ci/licenses_golden/licenses_skia | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)
```
what?!?

Trying again on 909b35818afdcf28ecb31df358e541965aad6b63 --- the first bad commit:
```
234 / 123
239 / 131
249 / 134
222 / 128
227 / 121
```
Does indeed seem bad.

Trying again on da2884a1c5bcc1cce2c6e726554f2036713c5c20 --- the commit before the first bad commit:
```
244 / 129
273 / 152
235 / 125
250 / 129
240 / 126
```
sooo that's not the 110-ish compile time we saw before... I'll try again with a completely clean build:

```
245 / 121
240 / 120
240 / 127
217 / 118
238 / 125
```
so da2884a1c5bcc1cce2c6e726554f2036713c5c20 is bad too --- and it seems that I can't trust non-clean builds.
In the below I thus do (slower) clean builds by removing the out directories before I compile: `rm -rf out/host_debug ; rm -rf out/android_debug_unopt_x64 ; ./flutter/tools/gn --no-prebuilt-dart-sdk ; ./flutter/tools/gn --no-prebuilt-dart-sdk --android --unoptimize --android-cpu x64 ;  ninja -C out/host_debug/ -j500 && ninja -C out/android_debug_unopt_x64/ -j500`

Rechecking fc733b35c5625d4b20c3511f58ea545f28c852a4 with a clean build:
```
213 / 102
206 / 110
218 / 109
211 / 107
214 / 110
```
 => that's actually good!

fe7861cafbfb7a1baa8675f4a2d1d0c2c36c86db on a clean build:
```
234 / 108
224 / 108
283 / 179
222 / 107
230 / 119
223 / 112
```
most are around 110 --- let's say good.

41f76722348853850f69a71e7567d354af9685b1 on a clean build:
```
268 / 122
259 / 123
225 / 121
223 / 118
225 / 121
```
most around 120 --- let's say it's bad.

That makes 41f76722348853850f69a71e7567d354af9685b1 the first bad one.
That commit bumps dart from aaca2ac128aeb3389a6aef41bd21832cfe632fae to 9f3cd7a49814ac36e85392c00a24a7c764c21852

doing a git log in `flutter/engine/src/third_party/dart` when it's on 9f3cd7a49814ac36e85392c00a24a7c764c21852 shows 
```
9f3cd7a4981 (HEAD) Version 2.15.0-222.0.dev
938e7ae54e4 [co19] Roll co19_2 to a93f8484f7265f0849692d3b1208323393733f8d
aaca2ac128a Version 2.15.0-221.0.dev
```

but looking at https://github.com/dart-lang/sdk/commits/dev?after=663d08ea8777580de1cf57309a7fed153e3317f8+69&branch=dev I see

```
9f3cd7a (9f3cd7a49814ac36e85392c00a24a7c764c21852)
7154635 (7154635d4ddf06d9dd548c55448a1d6a6029c3d9)
391e900 (391e900e5e2d67f73ccba0c78aef3781750e4d70)
66cbeb6 (66cbeb67188c69c340bcafd5f454608f693fea96)
72daafd (72daafd70c417dd5c790215e469affd21692ce29)
bb06e64 (bb06e649f46224e7d45b99efebce6e10a3a407d9)
ff26406 (ff26406d724e680f810e3b8d3709864ca470a145)
2bbb5dc (2bbb5dc9071f67aaf4808a1cc70c482ccbfafb7e) <- this is johnnis commit
938e7ae (938e7ae54e44f877042133258c58f4663512b6f8)
aaca2ac (aaca2ac128aeb3389a6aef41bd21832cfe632fae)
```

there's clearly something here I don't understand.
If someone knows more about git and/or github than I do please enlighten me.

Trying to check out dart at revision aaca2ac128a and doing a clean compile:
```
220 / 111
219 / 106
227 / 112
224 / 122
226 / 116
219 / 112
```
not so clear cut, but let's say good.

git diff aaca2ac128a...9f3cd7a49814ac36e85392c00a24a7c764c21852 shows the diff shown in https://github.com/dart-lang/sdk/commit/9f3cd7a49814ac36e85392c00a24a7c764c21852
Inspecting pkg/front_end/lib/src/fasta/incremental_compiler.dart shows that johnnis commit is not included. We can also see that at https://github.com/dart-lang/sdk/blob/9f3cd7a49814ac36e85392c00a24a7c764c21852/pkg/front_end/lib/src/fasta/incremental_compiler.dart#L371

Again on dart commit aaca2ac128aeb3389a6aef41bd21832cfe632fae:
```
224 / 110
232 / 112
212 / 108
215 / 106
216 / 106
```
Again it's good (and it's pretty clear cut :)).

Let's try to apply the diff from https://github.com/dart-lang/sdk/commit/9f3cd7a49814ac36e85392c00a24a7c764c21852 to aaca2ac128aeb3389a6aef41bd21832cfe632fae:
```
294 / 128
220 / 119
263 / 121
240 / 121
229 / 121
```
Bad.

My guess is that it's the vm changes in that patch that's causing this. Removing the vm changes (so aaca2ac128a with the diff from https://github.com/dart-lang/sdk/commit/9f3cd7a49814ac36e85392c00a24a7c764c21852 *except the 3 files in runtime*):
```
226 / 106
211 / 104
209 / 106
227 / 111
214 / 110
```
Good.

Lets try with aaca2ac128a with *only* the vm changes from 9f3cd7a49814ac36e85392c00a24a7c764c21852:
```
246 / 130
224 / 123
239 / 130
237 / 127
230 / 123
```
Bad.

Conclusion: The vm changes in https://github.com/dart-lang/sdk/commit/9f3cd7a49814ac36e85392c00a24a7c764c21852 bumps the compile time. The vm changes looks to be https://dart-review.googlesource.com/c/sdk/+/216687.
